src/tool/hpcrun/sample-sources/gpu/gpu-blame/blame.c

```diff
@@ -0,0 +1,484 @@
+
+static queue_node_t*
+queue_node_alloc_helper
+(
+ queue_node_t **free_list
+)
+{
+  queue_node_t *first = *free_list;
+
+  if (first) {
+    *free_list = atomic_load(&first->next);
+  } else {
+    first = (queue_node_t *) hpcrun_malloc_safe(sizeof(queue_node_t));
+  }
+
+  memset(first, 0, sizeof(queue_node_t));
+  return first;
+}
+
+
+static void
+queue_node_free_helper
+(
+ queue_node_t **free_list,
+ queue_node_t *node
+)
+{
+  atomic_store(&node->next, *free_list);
+  *free_list = node;
+}
+
+
+static kernel_node_t*
+kernel_node_alloc_helper
+(
+ kernel_node_t **free_list
+)
+{
+  kernel_node_t *first = *free_list;
+
+  if (first) {
+    *free_list = atomic_load(&first->next);
+  } else {
+    first = (kernel_node_t *) hpcrun_malloc_safe(sizeof(kernel_node_t));
+  }
+  memset(first, 0, sizeof(kernel_node_t));
+  return first;
+}
+
+
+static void
+kernel_node_free_helper
+(
+ kernel_node_t **free_list,
+ kernel_node_t *node
+)
+{
+  atomic_store(&node->next, *free_list);
+  *free_list = node;
+}
+
+
+static void
+create_activity_object
+(
+ gpu_activity_t *ga,
+ cct_node_t *cct_node,
+ gpu_blame_shift_t *activity
+)
+{
+  ga->kind = GPU_ACTIVITY_BLAME_SHIFT;
+  ga->cct_node = cct_node;
+  ga->details.blame_shift.cpu_idle_time = activity->cpu_idle_time;
+  ga->details.blame_shift.gpu_idle_time = activity->gpu_idle_time;
+  ga->details.blame_shift.cpu_idle_cause_time = activity->cpu_idle_cause_time;
+}
+
+
+static void
+record_blame_shift_metrics
+(
+ cct_node_t *cct_node,
+ gpu_blame_shift_t *bs
+)
+{
+  gpu_activity_t ga;
+  create_activity_object(&ga, cct_node, bs);
+  gpu_metrics_attribute(&ga);
+}
+
+
+static void
+add_queue_node_to_splay_tree
+(
+ uint64_t queue_id
+)
+{
+  queue_node_t *node = queue_node_alloc_helper(&queue_node_free_list);
+  atomic_init(&node->next, NULL);
+  queue_map_insert(queue_id, node);
+}
+
+
+// Insert a new node in kernel_map corresponding to a kernel
+static void
+create_and_insert_kernel_entry
+(
+ uint64_t kernelexec_id,
+ cct_node_t *launcher_cct
+)
+{
+  kernel_node_t *kernel_node = kernel_node_alloc_helper(&kernel_node_free_list);
+  kernel_node->kernel_id = kernelexec_id;
+  kernel_node->launcher_cct = launcher_cct;
+  atomic_init(&kernel_node->next, NULL);
+  kernel_map_insert(kernelexec_id, kernel_node);
+}
+
+
+static void
+record_latest_kernel_end_time
+(
+ unsigned long kernel_end_time
+)
+{
+  unsigned long expected = atomic_load(&latest_kernel_end_time);
+  if (kernel_end_time > expected) {
+    atomic_compare_exchange_strong(&latest_kernel_end_time, &expected, kernel_end_time);
+  }
+}
+
+
+static void
+add_kernel_to_completed_list
+(
+ kernel_node_t *kernel_node
+)
+{
+  while (true) {
+    kernel_node_t *current_head = atomic_load(&completed_kernel_list_head);
+    atomic_store(&kernel_node->next, current_head);
+    if (atomic_compare_exchange_strong(&completed_kernel_list_head, &current_head, kernel_node)) {
+      break;
```

Comment:
The above code is confusing to me. 

If you just want to append a node atomically to a concurrent list's head, you can follow the simple pattern below. No need to use an additional atomic load for list head.

https://github.com/HPCToolkit/hpctoolkit/blob/d53f6fce018673bc886305330ca91b686844468e/src/lib/prof-lean/stacks.c#L230

Reply:
@Jokeren , if the CAS in the while loop evaluates to false, where is the head variable getting updated?

Reply:
The head variable is updated by `atomic_compare_exchange_strong`